implicit/cuda/bpr.cu

```diff
@@ -0,0 +1,121 @@
+#include <stdio.h>
+#include <math.h>
+#include <cuda_runtime.h>
+#include <curand.h>
+#include <cublas_v2.h>
+
+
+#include "implicit/cuda/als.h"
+#include "implicit/cuda/utils.cuh"
+
+namespace implicit {
+__global__ void bpr_update_kernel(int samples, unsigned int * random_likes, unsigned int * random_dislikes,
```

Comment:
I really, really love this.

It's amazing to me that Hogwild works well even with such massive parallelism! How many parallel SGD updates does this actually do on a relatively good GPU?

Also, I _really_ need to get a GPU and get started on adding this to LightFM :)

Reply:
Thanks! 

On my GPU there are 3584 cuda cores - and each one is running a thread in parallel. 

I found CUDA programming relatively straightforward once I really wrapped my head around SIMT processing with warps of threads running the exact same instruction at the same time. I think that this CUDA code is even relatively easy to follow along with (though compared to lightfm it's a pretty simple model, I'm using SGD instead of adadelta, doesn't support item/user side features etc).

I just read your mixture paper btw - nice work! I really like the idea of treating users as a mixture of interests, and I'm going to forward it on to a couple people that I think will benefit from using that approach.